src/extensions/default/NavigationAndHistory/NavigationProvider.js

```diff
@@ -0,0 +1,553 @@
+        jumpInProgress = true;
+        
+        // To ensure we don't reopen the same doc in the last known pane
+        // rather bring it to the same pane where user has opened it
+        var thisDoc = DocumentManager.getOpenDocumentForPath(this.file);
+        if (thisDoc && thisDoc._masterEditor) {
+            this.paneId = thisDoc._masterEditor._paneId;
+        }
+        
+        CommandManager.execute(Commands.FILE_OPEN, {fullPath: this.file, paneId: this.paneId}).done(function () {
+            EditorManager.getCurrentFullEditor().setSelections(self.selections, true);
+            _validateNavigationCmds();
+        }).always(function () {
+            jumpInProgress = false;
+        });
+    };
+    
+    
+   /**
+    * Function to capture a non-zero set of selections as a navigation frame.
+    * The assumptions behind capturing a frame as a navigation frame are :
+    *
+    * -> If it's set by user explicitly (using mouse click or jump to definition)
+    * -> By clicking on search results
+    * -> Change of cursor by keyboard navigation keys or actual edits are not captured.
+    *
+    * @private
+    */
+    function _recordJumpDef(event, selectionObj) {
+        // Don't capture frames if we are navigating or document text is being refreshed(fileSync in progress)
+        if (jumpInProgress || (event.target && event.target.document._refreshInProgress)) {
+            return;
+        }
+        // Reset forward navigation stack if we are capturing a new event
+        jumpedPosStack = [];
+        if (captureTimer) {
+            window.clearTimeout(captureTimer);
+            captureTimer = null;
+        }
+        
+        // Ensure cursor activity has not happened because of arrow keys or edit
+        if (selectionObj.origin !== "+move" && (!window.event || (window.event && window.event.type !== "input"))) {
+            captureTimer = window.setTimeout(function () {
+                // Check if we have reached MAX_NAV_FRAMES_COUNT
+                // If yes, control overflow
+                if (jumpToPosStack.length === MAX_NAV_FRAMES_COUNT) {
+                    var navFrame = jumpToPosStack.splice(0, 1)[0];
+                    navFrame._clearMarkers();
+                }
+
+                currentEditPos = new NavigationFrame(event.target, selectionObj);
+                jumpToPosStack.push(currentEditPos);
+                _validateNavigationCmds();
+                activePosNotSynced = false;
+            }, NAV_FRAME_CAPTURE_LATENCY);
+        } else {
+            activePosNotSynced = true;
+        }
+    }
+    
+   /**
+    * Command handler to navigate backward
+    */
+    function _navigateBack() {
+        if (!jumpedPosStack.length) {
+            if (activePosNotSynced) {
+                currentEditPos = new NavigationFrame(EditorManager.getCurrentFullEditor(), {ranges: EditorManager.getCurrentFullEditor()._codeMirror.listSelections()});
+                jumpedPosStack.push(currentEditPos);
```

Comment:
Suggestion:
Make the anonymous function inside `setTimeout` a "real" function and use it here, too.